### PR TITLE
[DistSQL]Support `show table metadata` syntax.

### PR DESCRIPTION
**Pull request by lanchengx**
image

### CODE REVIEW

File: shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/text/distsql/ral/common/show/executor/ShowTableMetadataExecutor.java [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+                .collect(Collectors.toCollection(LinkedList::new));
+        LinkedList<List<Object>> indexRows = schema.getTables().get(tableName).getIndexes().values().stream().map(each -> each.getName())
+                .map(each -> buildRow(schemaName, tableName, "INDEX", each)).collect(Collectors.toCollection(LinkedList::new));
+        result.addAll(columnRows);
+        result.addAll(indexRows);
+        return result;
+    }
+    
+    private List<Object> buildRow(final String schemaName, final String tableName, final String type, final String name) {
+        return Stream.of(schemaName, tableName, type, name).map(each -> (Object) each).collect(Collectors.toCollection(LinkedList::new));
```
Comment (TeslaCN) [inline shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/text/distsql/ral/common/show/executor/ShowTableMetadataExecutor.java:94]: How about `new ArrayList<>(Arrays.asList(schemaName, tableName, type, name))`?

File: shardingsphere-distsql/shardingsphere-distsql-parser/src/main/java/org/apache/shardingsphere/distsql/parser/core/common/CommonDistSQLStatementVisitor.java [hunk 2]
```diff
@@ -96,6 +99,12 @@ public ASTNode visitAlterResource(final AlterResourceContext ctx) {
         return new AlterResourceStatement(ctx.dataSource().stream().map(each -> (DataSourceSegment) visit(each)).collect(Collectors.toList()));
     }
     
+    @Override
+    public ASTNode visitShowTableMetadata(final ShowTableMetadataContext ctx) {
+        Set<String> tableNames = ctx.tableName().stream().map(each -> getIdentifierValue(each)).collect(Collectors.toSet());
```
Comment (RaigorJiang) [inline shardingsphere-distsql/shardingsphere-distsql-parser/src/main/java/org/apache/shardingsphere/distsql/parser/core/common/CommonDistSQLStatementVisitor.java:104]: Please change the type definition of `tableNames` to Collection<String>.

File: shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/text/distsql/ral/common/show/executor/ShowTableMetadataExecutor.java [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    protected MergedResult createMergedResult() {
+        String schemaName = sqlStatement.getSchema().isPresent() ? sqlStatement.getSchema().get().getIdentifier().getValue() : connectionSession.getSchemaName();
+        if (null == schemaName) {
+            throw new NoDatabaseSelectedException();
+        }
+        if (!ProxyContext.getInstance().getAllSchemaNames().contains(schemaName)) {
+            throw new SchemaNotExistedException(schemaName);
+        }
+        ShardingSphereSchema schema = ProxyContext.getInstance().getMetaData(schemaName).getSchema();
+        LinkedList<List<Object>> rows = schema.getAllTableNames().stream().filter(each -> sqlStatement.getTableNames().contains(each))
```
Comment (RaigorJiang) [inline shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/text/distsql/ral/common/show/executor/ShowTableMetadataExecutor.java:77]: LinkedList -> Collection

File: shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/text/distsql/ral/common/show/executor/ShowTableMetadataExecutor.java [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+                .map(each -> buildTableRows(schemaName, schema, each)).flatMap(Collection::stream).collect(Collectors.toCollection(LinkedList::new));
+        return new MultipleLocalDataMergedResult(rows);
+    }
+    
+    private Collection<List<Object>> buildTableRows(final String schemaName, final ShardingSphereSchema schema, final String tableName) {
+        Collection<List<Object>> result = new LinkedList<>();
+        LinkedList<List<Object>> columnRows = schema.getAllColumnNames(tableName).stream().map(each -> buildRow(schemaName, tableName, "TABLE", each))
```
Comment (RaigorJiang) [inline shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/text/distsql/ral/common/show/executor/ShowTableMetadataExecutor.java:84]: LinkedList -> Collection

**Review by RaigorJiang — APPROVED** [review]

File: shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/text/distsql/ral/common/show/executor/ShowTableMetadataExecutor.java [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+        }
+        ShardingSphereSchema schema = ProxyContext.getInstance().getMetaData(schemaName).getSchema();
+        Collection<List<Object>> rows = schema.getAllTableNames().stream().filter(each -> sqlStatement.getTableNames().contains(each))
+                .map(each -> buildTableRows(schemaName, schema, each)).flatMap(Collection::stream).collect(Collectors.toCollection(LinkedList::new));
+        return new MultipleLocalDataMergedResult(rows);
+    }
+    
+    private Collection<List<Object>> buildTableRows(final String schemaName, final ShardingSphereSchema schema, final String tableName) {
+        Collection<List<Object>> result = new LinkedList<>();
+        Collection<List<Object>> columnRows = schema.getAllColumnNames(tableName).stream().map(each -> buildRow(schemaName, tableName, "TABLE", each))
```
Comment (RaigorJiang) [inline shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/text/distsql/ral/common/show/executor/ShowTableMetadataExecutor.java:84]: "TABLE" should be "COLUMN"

**Pull request merged**